zpcictl/zpcictl.8
.\" Copyright IBM Corp. 2022
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\"
.\" Macro for inserting an option description prologue.
.\" .OD <long> [<short>] [args]
.de OD
.  ds args "
.  if !'\\$3'' .as args \fI\\$3\fP
.  if !'\\$4'' .as args \\$4
.  if !'\\$5'' .as args \fI\\$5\fP
.  if !'\\$6'' .as args \\$6
.  if !'\\$7'' .as args \fI\\$7\fP
.  PD 0
.  if !'\\$2'' .IP "\fB\-\\$2\fP \\*[args]" 4
.  if !'\\$1'' .IP "\fB\-\-\\$1\fP \\*[args]" 4
.  PD
..
.
.TH zpcictl 8 "Mar 2022" s390-tools zpcictl
.
.SH NAME
zpcictl - Manage PCI devices on IBM Z
.
.
.SH SYNOPSIS
.B "zpcictl"
.I "OPTIONS"
.I "DEVICE"
.
.
.SH DESCRIPTION
Use
.B zpcictl
to manage PCI devices on the IBM Z platform. In particular,
use this command to report defective PCI devices to the Support Element (SE).

.B Note:
For NVMe devices additional data (such as S.M.A.R.T. data) is collected and sent
with any error handling action. For this extendend data collection, the
smartmontools must be installed.
.PP
.
.
.SH DEVICE
A PCI slot address (e.g. 0000:00:00.0) or the main device node of an NVMe
device (e.g.
.I /dev/nvme0
).
.
.
.SH OPTIONS
.SS Error Handling Options
.OD reset "" "DEVICE"
Reset and re-initialize the PCI device and report a device error to the Support
Element (SE). The reset consists of a controlled shutdown and a subsequent
re-enabling of the device from the shut off state. This process destroys and
then re-creates higher level interfaces such as network interfaces and block
devices. This reset is disruptive and often requires manual intervention on
multiple layers. In particular, network interfaces that are part of a bonded
interface must be re-added to the bond after the reset. Similarly, block
devices backed by an NVMe that are part of a software RAID must be re-synced by
re-adding to the RAID after resetting the NVMe.

Use this reset option only if the less disruptive automatic recovery mechanism
is not supported by your kernel or it failed to restore the device's
functionality. Unsuccessful automatic recovery can result in kernel messages
indicating required manual intervention. If the device is malfunctioning
without automatic recovery being triggered, consider using the \fB--reset-fw\fR
option to trigger a less disruptive automatic recovery through
a firmware-driven reset.
.PP
.
.OD reset-fw "" "DEVICE"
Reset the PCI device using a firmware-driven reset that also reports a device
error on the Support Element (SE). If supported by your kernel, automatic recovery
re-initializes the device after the firmware reports a successful device reset.

Use this option if the device is malfunctioning and automatic recovery is
supported by the kernel but was not triggered. This condition can occur if the
error is not detected by the low level PCI interfaces. A successful automatic
recovery after the firmware-driven reset, is less disruptive than the full
reset that is performed by the \fB--reset\fR option. Other than the full reset,
the automatic recovery does not completely shut down the device and re-create
it from the shut down state. Instead, it works with the device driver to
restore the device in place. Thus, higher level interfaces such as network
interfaces and block devices remain intact. In particular, with this type of
reset high availability mechanisms like a bonded network interface or
a software RAID can transparently re-integrate the recovered device. For
example, after a failure and recovery, a software RAID can resync a storage
device or a network interface can be re-integrated in a bond. In contrast to
a complete shut down, the device driver remains active and informs higher
layers of both the occurrence of an error state and the eventual recovery.
.PP
.
.OD deconfigure "" "DEVICE"
Deconfigure the PCI device and prepare for any repair action. This action
changes the status of the PCI device from configured to reserved.
.PP
.
.OD report-error "" "DEVICE"
Report any device error for the PCI device.
The device is marked as defective but no further action is taken.
.PP
.
.SS General Options
.OD help "h" ""
Print usage information, then exit.
.PP
.
.OD version "v" ""
Print version information, then exit.
.PP


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------

* __v2.xx.x (20xx-xx-xx)__

  For Linux kernel version: 6.x

  Add new tools / libraries:

  Changes of existing tools:

  Bug Fixes:

* __v2.37.0 (2025-02-07)__

  For Linux kernel version: 6.13

  Changes of existing tools:
  - dbginfo.sh: Add details on CPU-measurement
  - dbginfo.sh: Add new crypto command
  - dbginfo.sh: Add overview commands and crypto update
  - dbginfo.sh: Adding kdump info
  - dbginfo.sh: Removing outdated email references
  - dbginfo.sh: Rework network section
  - dbginfo.sh: Update copyright 2nd year
  - pvimg: Add '--(enable|disable)-image-encryption' flags to 'pvimg create'
  - pvimg: Add '--cck <FILE>' command line option and make '--comm-key' an alias
  - pvimg: Add '--hdr-key' command line option to 'pvimg create'
  - pvimg: Rename '--key' into '--hdr-key' and use '--key' as an alias (for 'pvimg info')
  - pvsecret: Add support for retrievable secrets
  - ziorep_config: Add PCHID field to adapter report
  - ziorep_traffic: Add DEVBUSID column to traffic report
  - ziorep_utilization: Add --fcp-device parameter to print virtual adapter report
  - ziorep_utilization: Add PCHID column to physical adapter report
  - ziorep_utilization: Now prints only physical adapter report by default
  - ziorep_utilization: Swap Bus-ID and CHPID columns in virtual adapter report
  - zipl/boot: Increase section size for eckd_mv dumper
  - zkey: Add support for listing and importing protected virtualization secrets

  Bug Fixes:
  - chpstat: Fix invalid utilization data on older kernels
  - opticsmon: Fix runaway loop in on_link_change()
  - zipl: Update inline assembly for GCC 15
  - zipl_helper.device-mapper: Add missed step in logical device resolution

* __v2.36.0 (2024-12-06)__

  For Linux kernel version: 6.12

  s390-tools: Define Rust MSRV as 1.75.0

  Add new tools / libraries:
  - cpacfinfo: Tool to provide CPACF information
  - opticsmon: Tools to monitor optical modules for directly attached PCI based NICs
  - pvimg: Rust rewrite of genprotimg

  Changes of existing tools:
  - chpstat: Add data bandwidth utilization column
  - chpstat: Add support for full CMCB
  - chpstat: Add support for new CMG types
  - dbginfo.sh: add overview commands and crypto update
  - genprotimg:
    1. genprotimg is now a symbolic link to the new tool `pvimg create`
    2. Breaking API changes in genprotimg:
       1. An existing output file is no longer silently overwritten; to revert
          to the old behavior use the `--overwrite` option.
       2. The Linux kernel component is now checked to determine whether it is a
          binary s390x kernel. To disable this behavior use the
          `--no-component-check` option.
  - hyptop: Support for structured output (json, json-seq, csv)
  - lszfcp: Add missing fallback marker for non-good fc_host port_state
  - lszfcp: Improve speed with many SCSI devices
  - pvattest: Add attestation policy check command
  - zipl: Add support of partitions of mirror md-devices

  Bug Fixes:
  - lszcrypt: Fix wrong state showing up for removed AP queue within SE guest
  - lszfcp: Show device names line for zfcp_units without SCSI device

* __v2.35.0 (2024-10-01)__

  For Linux kernel version: 6.11

  Add new tools / libraries:

  Changes of existing tools:
  - cpacfstats: Add support for FULL XTS (MSA 10) and HMAC (MSA 11) PAI counter
  - cpuplugd: Make cpuplugd compatible with hiperdispatch
  - dbginfo.sh: Add network sockstat info
  - pvapconfig: s390x exclusive build
  - zdev: Add option to select IPL device
  - zdump/dfo_s390: Support s390 DFO for vr-kernel dumps
  - zipl: Add support of mirror devices

  Bug Fixes:
  - (genprotimg|zipl)/boot: discard .note.package ELF section to save memory
  - netboot/mk-s390image: Fix size when argument is a symlink
  - ziorep_config: Fix warning message when multipath device is not there.
  - zipl: Fix problems when target parameters are specified by user
  - zipl: Fix segfault when creating device-based dumps with '--dry-run'

* __v2.34.0 (2024-08-01)__

  For Linux kernel version: 6.10

  Changes of existing tools:
  - ap_tools/ap-check: Add support for vfio-ap dynamic configuration
  - dbginfo.sh: Update/Add additional DASD data collection
  - dumpconf: Add new parameter 'SCP_DATA' for SCSI/NVMe/ECKD dump devices
  - libutil: Make formatted meta-data configurable
  - s390-tools: Replace 'which' with built-in 'command -v'
  - zdump/dfi_elf: Support core dumps of vr-kernels

  Bug Fixes:
  - chzdev: Fix warning about failed ATTR writes by udev
  - rust/pv: Try again if first CRL-URI is invalid
  - rust/pvattest: Add short option for --arpk
  - zdump: Fix 'zgetdump -i' ioctl error on s390 formatted dump file

* __v2.33.1 (2024-05-28)__

  For Linux kernel version: 6.9

  Bug Fixes:
  - s390-tools: Fix formatting and typos in README.md
  - s390-tools: Fix release string

* __v2.33.0 (2024-05-27)__

  For Linux kernel version: 6.9

  Add new tools / libraries:
  - chpstat: New tool for displaying channel path statistics
  - libutil: Add output format helpers(util_fmt: JSON, JSON-SEQ, CSV, text pairs)

  Changes of existing tools / libraries:
  - chzdev: Add --is-owner to identify files created by zdev
  - dasdfmt: Change default mode to always use full-format (Note: affects ESE DASD)
  - libap: Significantly reduce delay time between file lock retries
  - pvattest: Rewrite from C to Rust
  - pvattest: Support additional data & user-data
  - rust/pv: Support for Attestation

  Bug Fixes:
  - chreipl: Improve disk type detection when running under QEMU
  - dbginfo.sh: Use POSIX option with uname
  - s390-tools: Fix missing hyphen escapes in the man page for many tools
  - zipl/src: Fix bugs in disk_get_info() reproducible in corner cases

* __v2.32.0 (2024-04-03)__

  For Linux kernel version: 6.8

  Changes of existing tools:
  - cpumf/lscpumf: add support for machine type 3932

zpcictl/zpcictl.c
/*
 * zpcictl - Manage PCI devices on z Systems
 *
 * Copyright IBM Corp. 2018
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

#include <errno.h>
#include <fcntl.h>
#include <stddef.h>
#include <sys/stat.h>
#include <sys/sysmacros.h>
#include <time.h>

#include "lib/util_base.h"
#include "lib/util_libc.h"
#include "lib/util_opt.h"
#include "lib/util_path.h"
#include "lib/util_prg.h"
#include "lib/util_proc.h"
#include "lib/util_rec.h"
#include "lib/util_scandir.h"
#include "lib/util_sys.h"

#include "zpcictl.h"

#define SMARTCTL_CMDLINE "smartctl -x %s 2>/dev/null"

static const struct util_prg prg = {
	.desc = "Use zpcictl to manage PCI devices on IBM Z\n"
		"DEVICE is the slot ID or node of the device "
		"(e.g. 0000:00:00.0 or /dev/nvme0)",
	.args = "DEVICE",
	.copyright_vec = {
		{
			.owner = "IBM Corp.",
			.pub_first = 2018,
			.pub_last = 2018,
		},
		UTIL_PRG_COPYRIGHT_END
	}
};

/* Defines for options with no short command */
#define OPT_RESET	128
#define OPT_DECONF	129
#define OPT_REPORT_ERR	130
#define OPT_RESET_FW	131

static struct util_opt opt_vec[] = {
	UTIL_OPT_SECTION("ERROR HANDLING OPTIONS"),
	{
		.option = { "reset", no_argument, NULL, OPT_RESET },
		.desc = "Reset the device and report an error to the Support Element (SE). "
			"The reset consists of a controlled shutdown and a subsequent "
			"re-enabling of the device. As a result, higher level interfaces such "
			"as network interfaces and block devices are destroyed and re-created.\n"
			"Manual configuration steps might be required to re-integrate the device, "
			"for example, in bonded interfaces or software RAIDs.\n"
			"Use this option only if the automatic recovery failed, or if it did "
			"not succeed to restore regular operations of the device and manual "
			"intervention is required.\n",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},
	{
		.option = { "reset-fw", no_argument, NULL, OPT_RESET_FW },
		.desc = "Reset the device through a firmware driven reset that triggers "
			"automatic recovery and reports an error to the Support Element (SE).\n",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},
	{
		.option = { "deconfigure", no_argument, NULL, OPT_DECONF },
		.desc = "Deconfigure the device to prepare for any repair action",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},
	{
		.option = { "report-error", no_argument, NULL, OPT_REPORT_ERR },
		.desc = "Report a device error to the Support Element (SE)",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},
	UTIL_OPT_SECTION("GENERAL OPTIONS"),
	UTIL_OPT_HELP,
	UTIL_OPT_VERSION,
	UTIL_OPT_END
};

static int is_char_dev(const char *dev)
{
	struct stat s;

	if (stat(dev, &s))
		return 0;

	return S_ISCHR(s.st_mode);
}

static int is_blk_dev(const char *dev)
{
	struct stat s;

	if (stat(dev, &s))
		return 0;

	return S_ISBLK(s.st_mode);
}

static void fopen_err(char *path)
{
	warnx("Could not open file %s: %s", path, strerror(errno));
	free(path);
	exit(EXIT_FAILURE);
}

static void fclose_err(char *path)
{
	if (errno == EIO || errno == EOPNOTSUPP)
		warnx("Unsupported operation: %s: %s", path, strerror(errno));
	else
		warnx("Could not close file: %s: %s", path, strerror(errno));
	free(path);
	exit(EXIT_FAILURE);

}

static void fread_err(FILE *fp, char *path)
{
	warnx("Could not read file: %s: %s", path, strerror(errno));
	if (fclose(fp))
		fclose_err(path);
	free(path);
	exit(EXIT_FAILURE);
}

static void fwrite_err(FILE *fp, char *path)
{
	warnx("Could not write to file: %s: %s", path, strerror(errno));
	if (fclose(fp))
		fclose_err(path);
	free(path);
	exit(EXIT_FAILURE);
}

#define READ_CHUNK_SIZE		512

static char *collect_smart_data(struct zpci_device *pdev)
{
	char *buffer = NULL;
	size_t count = 0;
	char *cmd;
	FILE *fd;

	if (!pdev->device)
		return NULL;

	util_asprintf(&cmd, SMARTCTL_CMDLINE, pdev->device);
	fd = popen(cmd, "r");
	if (!fd)
		goto out;

	while (!feof(fd)) {
		buffer = realloc(buffer, count + READ_CHUNK_SIZE);
		if (!buffer) {
			warnx("Could not collect S.M.A.R.T. data");
			goto out;
		}
		count += fread(&buffer[count], 1, READ_CHUNK_SIZE, fd);
		if (ferror(fd)) {
			free(buffer);
			buffer = NULL;
			goto out;
		}
	}

	buffer = realloc(buffer, count);
	if (!buffer && count > 0)
		warnx("Could not collect S.M.A.R.T. data");
	if (buffer)
		buffer[count] = '\0';

out:
	pclose(fd);
	free(cmd);

	return buffer;
}

static unsigned int sysfs_read_value(struct zpci_device *pdev, const char *attr)
{
	unsigned int val;
	char *path;
	FILE *fp;

	path = util_path_sysfs("bus/pci/devices/%s/%s", pdev->slot, attr);
	fp = fopen(path, "r");
	if (!fp)
		fopen_err(path);
	if (fscanf(fp, "%x", &val) != 1) {
		fread_err(fp, path);
	}
	if (fclose(fp))
		fclose_err(path);
	free(path);

	return val;
}

static void sysfs_write_value(struct zpci_device *pdev, const char *attr,
				unsigned int val)
{
	char *path;
	FILE *fp;

	path = util_path_sysfs("bus/pci/devices/%s/%s", pdev->slot, attr);
	fp = fopen(path, "w");
	if (!fp)
		fopen_err(path);
	if (fprintf(fp, "%x", val) < 0) {
		fwrite_err(fp, path);
	}
	if (fclose(fp))
		fclose_err(path);
	free(path);
}

static void sysfs_report_error(struct zpci_report_error *report, char *slot)
{
	size_t r_size;
	char *path;
	FILE *fp;

	r_size = sizeof(*report);

	path = util_path_sysfs("bus/pci/devices/%s/report_error", slot);
	fp = fopen(path, "w");
	if (!fp)
		fopen_err(path);
	if (fwrite(report, 1, r_size, fp) != r_size)
		fwrite_err(fp, path);
	if (fclose(fp))
		fclose_err(path);
	free(path);
}

static void get_device_node(struct zpci_device *pdev)
{
	struct dirent **de_vec;
	char *path, *dev;
	char slot[13];
	int count, i;

	path = util_path_sysfs("bus/pci/devices/%s/nvme", pdev->slot);
	count = util_scandir(&de_vec, alphasort, path, "nvme*");
	if (count == -1) {
		warnx("Could not read directory %s: %s", path, strerror(errno));
		free(path);
		return;
	}

	for (i = 0; i < count; i++) {
		util_asprintf(&dev, "/dev/%s", de_vec[i]->d_name);
		if (util_sys_get_dev_addr(dev, slot) != 0)
			continue;
		if (strcmp(slot, pdev->slot) == 0) {
			pdev->device = dev;
			break;
		}
	}

	util_scandir_free(de_vec, count);
	free(path);
}

static int device_exists(char *dev)
{
	char *path;
	int rc = 0;

	/* In case a device node is specified, this will be sufficiant */
	if (util_path_exists(dev) && !util_path_is_dir(dev))
		return 1;

	path = util_path_sysfs("bus/pci/devices/%s", dev);
	if (util_path_exists(path))
		rc = 1;
	free(path);

	return rc;
}

static void get_device_info(struct zpci_device *pdev, char *dev)
{
	if (!device_exists(dev))
		errx(EXIT_FAILURE, "Could not find device %s", dev);
	if (is_blk_dev(dev))
		errx(EXIT_FAILURE, "Unsupported device type %s", dev);
	if (is_char_dev(dev)) {
		if (util_sys_get_dev_addr(dev, pdev->slot) != 0)
			errx(EXIT_FAILURE,
			     "Could not determine slot address for %s", dev);
		pdev->device = dev;
	} else {
		strcpy(pdev->slot, dev);
	}

	pdev->class = sysfs_read_value(pdev, "class");
	pdev->fid = sysfs_read_value(pdev, "function_id");
	pdev->pchid = sysfs_read_value(pdev, "pchid");

	/*
	 * In case a slot address was specified, the device node for NVMe
	 * devices is still needed. Otherwise it won't be possible to collect
	 * S.M.A.R.T. data at a later point.
	 */
	if (!pdev->device && pdev->class == PCI_CLASS_NVME)
		get_device_node(pdev);
}

/*
 * Issue an SCLP Adapter Error Notification event with a specific action
 * qualifier.
 *
 * Collect additional information when possible (e.g. S.M.A.R.T. data for NVMe
 * devices).
 */
static void sclp_issue_action(struct zpci_device *pdev, int action)
{
	struct zpci_report_error report = {
		.header = { 0 },
		.data = { 0 }
	};
	char *sdata = NULL;

	report.header.version = 1;
	report.header.action = action;
	report.header.length = offsetof(struct zpci_report_error_data, log_data);
	report.data.timestamp = (__u64)time(NULL);
	report.data.err_log_id = 0x4713;

	if (pdev->class == PCI_CLASS_NVME)
		sdata = collect_smart_data(pdev);
	if (sdata) {
		report.header.length += util_strlcpy(report.data.log_data, sdata,
						     sizeof(report.data.log_data));
		free(sdata);
	}
	sysfs_report_error(&report, pdev->slot);
}

/*
 * Reset the PCI device and initiate a re-initialization.
 */
static void sclp_reset_device(struct zpci_device *pdev)
{
	sclp_issue_action(pdev, SCLP_ERRNOTIFY_AQ_REPORT_ERR);
	sysfs_write_value(pdev, "recover", 1);
}

/*
 * Reset the PCI device via firmware and let auto recovery handle
 * re-initialization
 */
static void sclp_reset_device_fw(struct zpci_device *pdev)
{
	sclp_issue_action(pdev, SCLP_ERRNOTIFY_AQ_RESET);
}

/*
 * De-Configure/repair PCI device. Moves the device from configured
 * to reserved state.
 */
static void sclp_deconfigure(struct zpci_device *pdev)
{
	sclp_issue_action(pdev, SCLP_ERRNOTIFY_AQ_DECONF);
}

/*
 * Report an error to the SE.
 */
static void sclp_report_error(struct zpci_device *pdev)
{
	sclp_issue_action(pdev, SCLP_ERRNOTIFY_AQ_REPORT_ERR);
}

static void parse_cmdline(int argc, char *argv[], struct options *opts)
{
	int cmd;

	util_prg_init(&prg);
	util_opt_init(opt_vec, NULL);

	do {
		cmd = util_opt_getopt_long(argc, argv);

		switch (cmd) {
		case OPT_RESET:
			opts->reset = 1;
			break;
		case OPT_RESET_FW:
			opts->reset_fw = 1;
			break;
		case OPT_DECONF:
			opts->deconfigure = 1;
			break;
		case OPT_REPORT_ERR:
			opts->report = 1;
			break;
		case 'h':
			util_prg_print_help();
			util_opt_print_help();
			exit(EXIT_SUCCESS);
		case 'v':
			util_prg_print_version();
			exit(EXIT_SUCCESS);
		case -1:
			/* End of options string */
			if (argc == 1) {
				errx(EXIT_FAILURE,
				     "Use '%s --help' for more information",
				     argv[0]);
			}
			break;
		}
	} while (cmd != -1);
}

int main(int argc, char *argv[])
{
	struct zpci_device pdev = { 0 };
	struct options opts = { 0 };

	parse_cmdline(argc, argv, &opts);

	if (optind >= argc)
		errx(EXIT_FAILURE, "No device specified");

	get_device_info(&pdev, argv[optind]);

	if (opts.reset)
		sclp_reset_device(&pdev);
	if (opts.reset_fw)
		sclp_reset_device_fw(&pdev);
	else if (opts.deconfigure)
		sclp_deconfigure(&pdev);
	else if (opts.report)
		sclp_report_error(&pdev);

	return 0;
}
